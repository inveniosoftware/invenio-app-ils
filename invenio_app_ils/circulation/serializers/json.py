# -*- coding: utf-8 -*-
#
# Copyright (C) 2018-2020 CERN.
#
# invenio-app-ils is free software; you can redistribute it and/or modify it
# under the terms of the MIT License; see LICENSE file for more details.

"""Loan JSON serializers."""

from invenio_records_rest.serializers.json import JSONSerializer

from invenio_app_ils.literature.serializers.custom_fields import \
    field_cover_metadata

from .custom_fields import (field_is_overdue, field_pickup_location,
                            field_transaction_location, field_transaction_user)


class LoanJSONSerializer(JSONSerializer):
    """Serialize loan."""

    def transform_record(self, pid, record, links_factory=None, **kwargs):
        """Transform record into an intermediate representation."""
        loan = super().transform_record(
            pid, record, links_factory=links_factory, **kwargs
        )
        field_is_overdue(loan["metadata"])
        field_pickup_location(loan["metadata"])
        field_transaction_location(loan["metadata"])
        field_transaction_user(loan["metadata"])
        field_cover_metadata(loan["metadata"].get("document", {}))
        return loan

    def transform_search_hit(
        self, pid, record_hit, links_factory=None, **kwargs
    ):
        """Transform search result hit into an intermediate representation."""
        hit = super().transform_search_hit(
            pid, record_hit, links_factory=links_factory, **kwargs
        )
        field_is_overdue(hit["metadata"])
        field_cover_metadata(hit["metadata"].get("document", {}))
        return hit
