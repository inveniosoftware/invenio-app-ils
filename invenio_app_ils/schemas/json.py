# -*- coding: utf-8 -*-
#
# Copyright (C) 2019 CERN.
#
# invenio-app-ils is free software; you can redistribute it and/or modify it
# under the terms of the MIT License; see LICENSE file for more details.

"""Invenio App ILS schemas."""
from invenio_rest.serializer import BaseSchema
from marshmallow import fields


class ILSRecordSchemaJSONV1(BaseSchema):
    """Schema for records v1 in JSON."""

    id = fields.Str(attribute='pid.pid_value')
    metadata = fields.Raw()
    links = fields.Raw()
    created = fields.Str()
    updated = fields.Str()
