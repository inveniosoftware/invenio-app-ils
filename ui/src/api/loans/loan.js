import { http, apiConfig } from '../base';
import { sessionManager } from '@authentication/services';
import { toShortDate } from '../date';
import { DateTime } from 'luxon';
import { serializer } from './serializer';
import isEmpty from 'lodash/isEmpty';
import { prepareDateQuery, prepareSumQuery } from '../utils';
import { generatePath } from 'react-router-dom';

const apiPaths = {
  checkout: '/circulation/loans/checkout',
  emailOverdue: '/circulation/loans/:loanPid/email-overdue',
  item: '/circulation/loans/:loanPid',
  list: '/circulation/loans/',
  request: '/circulation/loans/request',
  replaceItem: '/circulation/loans/:loanPid/replace-item',
};

const get = async loanPid => {
  const path = generatePath(apiPaths.item, { loanPid: loanPid });
  const response = await http.get(path);
  response.data = serializer.fromJSON(response.data);
  return response;
};

const doAction = async (
  url,
  documentPid,
  patronPid,
  { itemPid = null, cancelReason = null } = {}
) => {
  const currentUser = sessionManager.user;
  const payload = {
    document_pid: documentPid,
    patron_pid: patronPid,
    transaction_location_pid: currentUser.locationPid,
    transaction_user_pid: currentUser.id,
  };
  if (itemPid) {
    payload.item_pid = itemPid;
  }
  if (cancelReason) {
    payload.cancel_reason = cancelReason;
  }

  const response = await http.post(url, payload);
  response.data = serializer.fromJSON(response.data);
  return response;
};

const doRequest = async (
  documentPid,
  patronPid,
  {
    requestExpireDate = null,
    requestStartDate = null,
    deliveryMethod = null,
  } = {}
) => {
  const currentUser = sessionManager.user;
  const payload = {
    document_pid: documentPid,
    patron_pid: patronPid,
    transaction_location_pid: currentUser.locationPid,
    transaction_user_pid: currentUser.id,
  };

  if (requestStartDate) {
    payload.request_start_date = requestStartDate;
  }
  if (requestExpireDate) {
    payload.request_expire_date = requestExpireDate;
  }
  if (deliveryMethod) {
    payload.delivery = {
      method: deliveryMethod,
    };
  }

  const response = await http.post(apiPaths.request, payload);
  response.data = serializer.fromJSON(response.data);
  return response;
};

const doCheckout = async (
  documentPid,
  itemPid,
  patronPid,
  { startDate = null, endDate = null, force = false } = {}
) => {
  const currentUser = sessionManager.user;
  const payload = {
    document_pid: documentPid,
    item_pid: itemPid,
    patron_pid: patronPid,
    transaction_location_pid: currentUser.locationPid,
    transaction_user_pid: currentUser.id,
  };

  if (startDate) {
    payload.start_date = startDate;
  }
  if (endDate) {
    payload.end_date = endDate;
  }
  if (force) {
    payload.force = true;
  }

  const response = await http.post(apiPaths.checkout, payload);
  response.data = serializer.fromJSON(response.data);
  return response;
};

const assignItemToLoan = async (itemPid, loanPid) => {
  const path = generatePath(apiPaths.replaceItem, { loanPid: loanPid });
  const payload = { item_pid: itemPid };
  const response = await http.post(path, payload);
  response.data = serializer.fromJSON(response.data);
  return response;
};

class QueryBuilder {
  constructor() {
    this.documentQuery = [];
    this.itemQuery = [];
    this.overdueQuery = [];
    this.patronQuery = [];
    this.renewedCountQuery = [];
    this.size = '';
    this.page = '';
    this.sortBy = '';
    this.startDateQuery = [];
    this.stateQuery = [];
    this.updatedQuery = [];
  }

  withDocPid(documentPid) {
    if (
      !documentPid ||
      (typeof documentPid != 'number' && isEmpty(documentPid))
    ) {
      throw TypeError('DocumentPid argument missing');
    }
    this.documentQuery.push(`document_pid:${prepareSumQuery(documentPid)}`);
    return this;
  }

  withItemPid(itemPid) {
    if (!itemPid || (typeof itemPid != 'number' && isEmpty(itemPid))) {
      throw TypeError('itemPid argument missing');
    }
    this.itemQuery.push(`item_pid:${prepareSumQuery(itemPid)}`);
    return this;
  }

  withPatronPid(patronPid) {
    if (!patronPid || (typeof patronPid != 'number' && isEmpty(patronPid))) {
      throw TypeError('patronPid argument missing');
    }
    this.patronQuery.push(`patron_pid:${prepareSumQuery(patronPid)}`);
    return this;
  }

  withState(state) {
    if (!state || isEmpty(state)) {
      throw TypeError('state argument missing');
    }
    this.stateQuery.push(`state:${prepareSumQuery(state)}`);
    return this;
  }

  overdue() {
    let now = toShortDate(DateTime.local());
    this.overdueQuery.push(encodeURI(`end_date:{* TO ${now}}`));
    return this;
  }

  withUpdated(dates) {
    this.updatedQuery.push(
      prepareDateQuery('_updated', dates.date, dates.from, dates.to)
    );
    return this;
  }

  withStartDate({ fromDate, toDate }) {
    if (fromDate || toDate)
      this.startDateQuery.push(
        prepareDateQuery('start_date', null, fromDate, toDate)
      );
    return this;
  }

  /**
   * Combine elasticsearch query for number of renewals
   * @param renewals string, number or array
   * when number it asks for exact number
   * in string there is possibility of passing comparison operators
   */
  withRenewedCount(renewals) {
    if (
      !renewals ||
      isEmpty(renewals) ||
      !(typeof renewals === 'number' || typeof renewals === 'string')
    ) {
      throw TypeError('Renewal argument missing or invalid type');
    }
    this.renewedCountQuery.push(`extension_count:${renewals}`);
    return this;
  }

  withSize(size) {
    if (size > 0) this.size = `&size=${size}`;
    return this;
  }

  withPage(page = 0) {
    if (page > 0) this.page = `&page=${page}`;
    return this;
  }

  sortByNewest() {
    this.sortBy = `&sort=-mostrecent`;
    return this;
  }

  qs() {
    const searchCriteria = this.documentQuery
      .concat(
        this.itemQuery,
        this.patronQuery,
        this.stateQuery,
        this.overdueQuery,
        this.updatedQuery,
        this.renewedCountQuery,
        this.startDateQuery
      )
      .join(' AND ');
    return `(${searchCriteria})${this.sortBy}${this.size}${this.page}`;
  }
}

const queryBuilder = () => {
  return new QueryBuilder();
};

const list = async query => {
  const response = await http.get(`${apiPaths.list}?q=${query}`);
  response.data.total = response.data.hits.total;
  response.data.hits = response.data.hits.hits.map(hit =>
    serializer.fromJSON(hit)
  );
  return response;
};

const sendOverdueLoansMailReminder = async payload => {
  const path = generatePath(apiPaths.emailOverdue, {
    loanPid: payload.loanPid,
  });
  return await http.post(path, payload);
};

const count = async query => {
  const response = await http.get(`${apiPaths.list}?q=${query}`);
  response.data = response.data.hits.total;
  return response;
};

export const loan = {
  searchBaseURL: `${apiConfig.baseURL}${apiPaths.list}`,
  assignItemToLoan: assignItemToLoan,
  query: queryBuilder,
  list: list,
  get: get,
  count: count,
  doAction: doAction,
  doRequest: doRequest,
  doCheckout: doCheckout,
  sendOverdueLoansMailReminder: sendOverdueLoansMailReminder,
  serializer: serializer,
};
