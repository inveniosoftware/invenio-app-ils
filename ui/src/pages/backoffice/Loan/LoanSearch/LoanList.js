import { DocumentIcon, ItemIcon, LoanIcon } from '@pages/backoffice';
import React, { Component } from 'react';
import { Grid, Header, Icon, Item, Label, List } from 'semantic-ui-react';
import { Link } from 'react-router-dom';
import { DocumentAuthors } from '@components/Document';
import { SearchEmptyResults } from '@components/SearchControls';
import { BackOfficeRoutes, DetailsRouteByPidTypeFor } from '@routes/urls';
import { OverdueLoanSendMailModal } from '../../components/OverdueLoanSendMailModal';
import { invenioConfig } from '@config';

class LoanDates extends Component {
  render() {
    const { loan } = this.props;
    if (
      invenioConfig.circulation.loanRequestStates.includes(loan.metadata.state)
    ) {
      return (
        <>
          <List.Content floated="right">
            {loan.metadata.request_start_date}
          </List.Content>
          <List.Content>
            <label> Requested on </label>
          </List.Content>
          <List.Content floated="right">
            {loan.metadata.request_expire_date}
          </List.Content>
          <List.Content>
            <label> Expires on </label>
          </List.Content>
        </>
      );
    } else {
      return (
        <>
          <List.Content floated="right">
            {loan.metadata.start_date}
          </List.Content>
          <List.Content>
            <label> Start date </label>
          </List.Content>
          <List.Content floated="right">
            {loan.metadata.is_overdue && <Icon name="warning" />}
            {loan.metadata.end_date}
          </List.Content>
          <List.Content>
            <label> End date </label>
          </List.Content>
        </>
      );
    }
  }
}

class LoanListEntry extends Component {
  getLinkToItem = itemPid => {
    return (
      <Link to={DetailsRouteByPidTypeFor(itemPid.type)(itemPid.value)}>
        ${itemPid.value}
      </Link>
    );
  };

  render() {
    const { loan } = this.props;

    return (
      <Item>
        <Item.Content>
          {loan.metadata.is_overdue && (
            <Label color="red" ribbon>
              Overdue
            </Label>
          )}
          <Item.Header
            as={Link}
            to={BackOfficeRoutes.loanDetailsFor(loan.metadata.pid)}
            data-test={`navigate-${loan.metadata.pid}`}
          >
            <LoanIcon /> Loan #{loan.metadata.pid}
          </Item.Header>
          <Grid columns={5}>
            <Grid.Column computer={6} largeScreen={5}>
              <label>Patron</label>{' '}
              <Link
                to={BackOfficeRoutes.patronDetailsFor(loan.metadata.patron_pid)}
              >
                {loan.metadata.patron.name}
              </Link>{' '}
              requested:
              <Item.Meta className={'document-authors'}>
                <Header className="loan-document-title" as="h5">
                  {loan.metadata.document.title}
                </Header>
                <DocumentAuthors
                  metadata={loan.metadata.document}
                  prefix={'by '}
                />
              </Item.Meta>
            </Grid.Column>
            <Grid.Column computer={3} largeScreen={3}>
              <List>
                <List.Item>
                  <List.Content floated="right">
                    {loan.metadata.state}
                  </List.Content>
                  <List.Content>
                    <label>State</label>
                  </List.Content>
                </List.Item>
                <List.Item>
                  <LoanDates loan={loan} />
                </List.Item>
                <List.Item>
                  <List.Content floated="right">
                    {loan.metadata.extension_count || '0'}
                  </List.Content>
                  <List.Content>
                    <label> Extensions</label>
                  </List.Content>
                </List.Item>
              </List>
            </Grid.Column>
            <Grid.Column width={2}>
              <OverdueLoanSendMailModal loan={loan} />
            </Grid.Column>
            <Grid.Column computer={3} largeScreen={3}>
              {loan.metadata.item_pid && (
                <>
                  <List>
                    <List.Item>
                      <List.Content>
                        Item{' '}
                        <Link
                          to={DetailsRouteByPidTypeFor(
                            loan.metadata.item_pid.type
                          )(loan.metadata.item_pid.value)}
                        >
                          {loan.metadata.item.barcode && (
                            <>
                              <ItemIcon />
                              {loan.metadata.item.barcode}
                            </>
                          )}
                        </Link>
                      </List.Content>
                      {loan.metadata.item.medium && (
                        <List.Content>
                          <label>medium</label> {loan.metadata.item.medium}
                        </List.Content>
                      )}
                    </List.Item>
                  </List>
                </>
              )}
            </Grid.Column>
            <Grid.Column computer={2} largeScreen={2} textAlign="right">
              <Item.Meta className={'pid-field'}>
                <Header disabled as="h5" className={'pid-field'}>
                  #{loan.metadata.pid}
                </Header>
              </Item.Meta>
              <Link
                to={BackOfficeRoutes.documentDetailsFor(
                  loan.metadata.document_pid
                )}
              >
                Document <DocumentIcon />
              </Link>
              <br />
            </Grid.Column>
          </Grid>
        </Item.Content>
      </Item>
    );
  }
}

export default class LoanList extends Component {
  renderListEntry = loan => {
    if (this.props.renderListEntryElement) {
      return this.props.renderListEntryElement(loan);
    }
    return <LoanListEntry key={loan.metadata.pid} loan={loan} />;
  };

  render() {
    const { hits } = this.props;

    if (!hits.length) return <SearchEmptyResults />;

    return (
      <Item.Group divided className={'bo-loan-search'}>
        {hits.map(hit => {
          return this.renderListEntry(hit);
        })}
      </Item.Group>
    );
  }
}
