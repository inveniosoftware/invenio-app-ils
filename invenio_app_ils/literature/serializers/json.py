# -*- coding: utf-8 -*-
#
# Copyright (C) 2018-2020 CERN.
#
# invenio-app-ils is free software; you can redistribute it and/or modify it
# under the terms of the MIT License; see LICENSE file for more details.

"""Literature JSON serializers."""

from invenio_records_rest.serializers.json import JSONSerializer

from .custom_fields import field_cover_metadata


class LiteratureJSONSerializer(JSONSerializer):
    """Serialize Literature."""

    def transform_record(self, pid, record, links_factory=None, **kwargs):
        """Transform record into an intermediate representation."""
        literature = super().transform_record(
            pid, record, links_factory=links_factory, **kwargs
        )
        field_cover_metadata(literature["metadata"])
        return literature

    def transform_search_hit(
        self, pid, record_hit, links_factory=None, **kwargs
    ):
        """Transform search result hit into an intermediate representation."""
        hit = super().transform_search_hit(
            pid, record_hit, links_factory=links_factory, **kwargs
        )
        field_cover_metadata(hit["metadata"])
        return hit
