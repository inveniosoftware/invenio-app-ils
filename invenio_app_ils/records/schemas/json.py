# -*- coding: utf-8 -*-
#
# Copyright (C) 2019 CERN.
#
# invenio-app-ils is free software; you can redistribute it and/or modify it
# under the terms of the MIT License; see LICENSE file for more details.

"""Invenio App ILS schemas."""

from invenio_records_rest.schemas.json import RecordSchemaJSONV1
from marshmallow import fields


class ILSRecordSchemaJSONV1(RecordSchemaJSONV1):
    """Schema for records v1 in JSON."""

    class Meta:
        """Meta attributes for the schema."""

        from marshmallow import EXCLUDE
        unknown = EXCLUDE

    id = fields.Str(attribute='pid.pid_value')
    metadata = fields.Raw()
    links = fields.Raw()
    created = fields.Str()
    updated = fields.Str()
