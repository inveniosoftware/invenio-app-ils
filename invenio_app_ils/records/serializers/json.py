# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2016-2018 CERN.
#
# Invenio is free software; you can redistribute it and/or modify it
# under the terms of the MIT License; see LICENSE file for more details.

"""Marshmallow based JSON serializer for records."""

from __future__ import absolute_import, print_function

from flask import json, request
from invenio_records_rest.serializers.base import PreprocessorMixin, \
    SerializerMixinInterface
from invenio_records_rest.serializers.marshmallow import MarshmallowMixin


class JSONSerializerMixin(SerializerMixinInterface):
    """Mixin serializing records as JSON."""

    @staticmethod
    def _format_args():
        """Get JSON dump indentation and separates."""
        if request and request.args.get('prettyprint'):
            return dict(
                indent=2,
                separators=(', ', ': '),
            )
        else:
            return dict(
                indent=None,
                separators=(',', ':'),
            )

    def serialize(self, loans, errors, record_serializer=None,
                  links_factory=None,
                  **kwargs):
        """Serialize a single record and persistent identifier.

        :param loans: list of loans
        :param errors: list of errors.
        :param links_factory: Factory function for record links.
        """
        return json.dumps(
            dict(
                loans=[record_serializer.transform_record(
                    loan['loan_pid'], loan, links_factory,
                    **kwargs) for loan in loans],
                errors=[error for error in errors]
            ),
            **self._format_args())


class MultipleCheckoutJSONSerializer(JSONSerializerMixin,
                                     MarshmallowMixin, PreprocessorMixin):
    """Marshmallow based JSON serializer for records."""
