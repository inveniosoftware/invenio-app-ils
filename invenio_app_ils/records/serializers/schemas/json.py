# -*- coding: utf-8 -*-
#
# Copyright (C) 2018-2019 CERN.
#
# invenio-app-ils is free software; you can redistribute it and/or modify it
# under the terms of the MIT License; see LICENSE file for more details.

"""Ils marshmallow schemas module."""
from invenio_records_rest.schemas import RecordSchemaJSONV1
from marshmallow import Schema, fields


class MultipleCheckoutErrorSchemaJSONV1(Schema):
    """Schema for handling multiple checkout response."""

    item_barcode = fields.Str()
    error_msg = fields.Str()
    error_code = fields.Str()


class CustomMultipleLoansSchemaJSONV1(Schema):
    """Schema for list of records v1 in JSON."""

    loans = fields.List(fields.Nested(RecordSchemaJSONV1))
    errors = fields.List(fields.Nested(MultipleCheckoutErrorSchemaJSONV1))
