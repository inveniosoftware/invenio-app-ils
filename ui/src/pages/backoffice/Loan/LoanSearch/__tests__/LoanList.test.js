import React from 'react';
import { shallow, mount } from 'enzyme';
import { BrowserRouter } from 'react-router-dom';
import LoanList from '../LoanList';
import * as testData from '@testData/loans.json';


jest.mock('../../../components/OverdueLoanSendMailModal', () => {
  return {
    OverdueLoanSendMailModal: () => null,
  };
});

describe('LoanList tests', () => {
  let component;
  afterEach(() => {
    if (component) {
      component.unmount();
    }
  });


  const document = {
    metadata: {
      ...testData[0],
      document: { title: 'TEST' },
    },
  };

  it('should load the LoanList component', () => {
    const component = shallow(<LoanList hits={[document]} />);
    expect(component).toMatchSnapshot();
  });

  it('should show empty message', () => {
    const component = shallow(<LoanList hits={[]} />);
    expect(component).toMatchSnapshot();
    // const resultRows = component
    //   .find('TableRow')
    //   .filterWhere(element => element.prop('data-test') === firstResult.id);
    // expect(resultRows).toHaveLength(1);
  });

  it('should load the LoanList component', () => {
    const component = mount(
      <BrowserRouter>
        <LoanList hits={[document]} />
      </BrowserRouter>
    );
    expect(component).toMatchSnapshot();
  });
});
