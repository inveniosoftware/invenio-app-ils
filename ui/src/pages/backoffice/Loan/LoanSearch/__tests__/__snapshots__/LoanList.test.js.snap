// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoanList tests should load the LoanList component 1`] = `
<ItemGroup
  className="bo-loan-search"
  divided={true}
>
  <LoanListEntry
    key="loanid-1"
    loan={
      Object {
        "metadata": Object {
          "document": Object {
            "title": "TEST",
          },
          "document_pid": "docid-1",
          "end_date": "2018-07-28",
          "item_pid": Object {},
          "patron": Object {
            "email": "patron1@test.ch",
          },
          "patron_pid": "1",
          "pickup_location_pid": "1",
          "pid": "loanid-1",
          "request_expire_date": "2019-06-28",
          "request_start_date": "2018-06-28",
          "start_date": "2018-06-28",
          "state": "PENDING",
          "transaction_date": "2018-06-29",
          "transaction_location_pid": "1",
        },
      }
    }
  />
</ItemGroup>
`;

exports[`LoanList tests should load the LoanList component 2`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <LoanList
      hits={
        Array [
          Object {
            "metadata": Object {
              "document": Object {
                "title": "TEST",
              },
              "document_pid": "docid-1",
              "end_date": "2018-07-28",
              "item_pid": Object {},
              "patron": Object {
                "email": "patron1@test.ch",
              },
              "patron_pid": "1",
              "pickup_location_pid": "1",
              "pid": "loanid-1",
              "request_expire_date": "2019-06-28",
              "request_start_date": "2018-06-28",
              "start_date": "2018-06-28",
              "state": "PENDING",
              "transaction_date": "2018-06-29",
              "transaction_location_pid": "1",
            },
          },
        ]
      }
    >
      <ItemGroup
        className="bo-loan-search"
        divided={true}
      >
        <div
          className="ui divided items bo-loan-search"
        >
          <LoanListEntry
            key="loanid-1"
            loan={
              Object {
                "metadata": Object {
                  "document": Object {
                    "title": "TEST",
                  },
                  "document_pid": "docid-1",
                  "end_date": "2018-07-28",
                  "item_pid": Object {},
                  "patron": Object {
                    "email": "patron1@test.ch",
                  },
                  "patron_pid": "1",
                  "pickup_location_pid": "1",
                  "pid": "loanid-1",
                  "request_expire_date": "2019-06-28",
                  "request_start_date": "2018-06-28",
                  "start_date": "2018-06-28",
                  "state": "PENDING",
                  "transaction_date": "2018-06-29",
                  "transaction_location_pid": "1",
                },
              }
            }
          >
            <Item>
              <div
                className="item"
              >
                <ItemContent>
                  <div
                    className="content"
                  >
                    <ItemHeader
                      as={
                        Object {
                          "$$typeof": Symbol(react.forward_ref),
                          "displayName": "Link",
                          "propTypes": Object {
                            "innerRef": [Function],
                            "onClick": [Function],
                            "replace": [Function],
                            "target": [Function],
                            "to": [Function],
                          },
                          "render": [Function],
                        }
                      }
                      data-test="navigate-loanid-1"
                      to="/backoffice/loans/loanid-1"
                    >
                      <Link
                        className="header"
                        data-test="navigate-loanid-1"
                        to="/backoffice/loans/loanid-1"
                      >
                        <LinkAnchor
                          className="header"
                          data-test="navigate-loanid-1"
                          href="/backoffice/loans/loanid-1"
                          navigate={[Function]}
                        >
                          <a
                            className="header"
                            data-test="navigate-loanid-1"
                            href="/backoffice/loans/loanid-1"
                            onClick={[Function]}
                          >
                            <LoanIcon>
                              <Icon
                                as="i"
                                name="bookmark outline"
                              >
                                <i
                                  aria-hidden="true"
                                  className="bookmark outline icon"
                                  onClick={[Function]}
                                />
                              </Icon>
                            </LoanIcon>
                             Loan #
                            loanid-1
                          </a>
                        </LinkAnchor>
                      </Link>
                    </ItemHeader>
                    <Grid
                      columns={5}
                    >
                      <div
                        className="ui five column grid"
                      >
                        <GridColumn
                          computer={6}
                          largeScreen={5}
                        >
                          <div
                            className="six wide computer five wide large screen column"
                          >
                            <label>
                              Patron
                            </label>
                             
                            <Link
                              to="/backoffice/patrons/1"
                            >
                              <LinkAnchor
                                href="/backoffice/patrons/1"
                                navigate={[Function]}
                              >
                                <a
                                  href="/backoffice/patrons/1"
                                  onClick={[Function]}
                                />
                              </LinkAnchor>
                            </Link>
                             
                            requested:
                            <ItemMeta
                              className="document-authors"
                            >
                              <div
                                className="meta document-authors"
                              >
                                <Header
                                  as="h5"
                                  className="loan-document-title"
                                >
                                  <h5
                                    className="ui header loan-document-title"
                                  >
                                    TEST
                                  </h5>
                                </Header>
                                <DocumentAuthors
                                  allFields={false}
                                  authorsLimit={10}
                                  delimiter="; "
                                  metadata={
                                    Object {
                                      "title": "TEST",
                                    }
                                  }
                                  prefix="by "
                                  scrollLimit={Infinity}
                                >
                                  <div
                                    className="document-authors-list-wrapper undefined"
                                  >
                                    by  
                                  </div>
                                </DocumentAuthors>
                              </div>
                            </ItemMeta>
                          </div>
                        </GridColumn>
                        <GridColumn
                          computer={3}
                          largeScreen={3}
                        >
                          <div
                            className="three wide computer three wide large screen column"
                          >
                            <List>
                              <div
                                className="ui list"
                                role="list"
                              >
                                <ListItem>
                                  <div
                                    className="item"
                                    onClick={[Function]}
                                    role="listitem"
                                  >
                                    <ListContent
                                      floated="right"
                                    >
                                      <div
                                        className="right floated content"
                                      >
                                        PENDING
                                      </div>
                                    </ListContent>
                                    <ListContent>
                                      <div
                                        className="content"
                                      >
                                        <label>
                                          State
                                        </label>
                                      </div>
                                    </ListContent>
                                  </div>
                                </ListItem>
                                <ListItem>
                                  <div
                                    className="item"
                                    onClick={[Function]}
                                    role="listitem"
                                  >
                                    <LoanDates
                                      loan={
                                        Object {
                                          "metadata": Object {
                                            "document": Object {
                                              "title": "TEST",
                                            },
                                            "document_pid": "docid-1",
                                            "end_date": "2018-07-28",
                                            "item_pid": Object {},
                                            "patron": Object {
                                              "email": "patron1@test.ch",
                                            },
                                            "patron_pid": "1",
                                            "pickup_location_pid": "1",
                                            "pid": "loanid-1",
                                            "request_expire_date": "2019-06-28",
                                            "request_start_date": "2018-06-28",
                                            "start_date": "2018-06-28",
                                            "state": "PENDING",
                                            "transaction_date": "2018-06-29",
                                            "transaction_location_pid": "1",
                                          },
                                        }
                                      }
                                    >
                                      <ListContent
                                        floated="right"
                                      >
                                        <div
                                          className="right floated content"
                                        >
                                          2018-06-28
                                        </div>
                                      </ListContent>
                                      <ListContent>
                                        <div
                                          className="content"
                                        >
                                          <label>
                                             Requested on 
                                          </label>
                                        </div>
                                      </ListContent>
                                      <ListContent
                                        floated="right"
                                      >
                                        <div
                                          className="right floated content"
                                        >
                                          2019-06-28
                                        </div>
                                      </ListContent>
                                      <ListContent>
                                        <div
                                          className="content"
                                        >
                                          <label>
                                             Expires on 
                                          </label>
                                        </div>
                                      </ListContent>
                                    </LoanDates>
                                  </div>
                                </ListItem>
                                <ListItem>
                                  <div
                                    className="item"
                                    onClick={[Function]}
                                    role="listitem"
                                  >
                                    <ListContent
                                      floated="right"
                                    >
                                      <div
                                        className="right floated content"
                                      >
                                        0
                                      </div>
                                    </ListContent>
                                    <ListContent>
                                      <div
                                        className="content"
                                      >
                                        <label>
                                           Extensions
                                        </label>
                                      </div>
                                    </ListContent>
                                  </div>
                                </ListItem>
                              </div>
                            </List>
                          </div>
                        </GridColumn>
                        <GridColumn
                          width={2}
                        >
                          <div
                            className="two wide column"
                          >
                            <OverdueLoanSendMailModal
                              loan={
                                Object {
                                  "metadata": Object {
                                    "document": Object {
                                      "title": "TEST",
                                    },
                                    "document_pid": "docid-1",
                                    "end_date": "2018-07-28",
                                    "item_pid": Object {},
                                    "patron": Object {
                                      "email": "patron1@test.ch",
                                    },
                                    "patron_pid": "1",
                                    "pickup_location_pid": "1",
                                    "pid": "loanid-1",
                                    "request_expire_date": "2019-06-28",
                                    "request_start_date": "2018-06-28",
                                    "start_date": "2018-06-28",
                                    "state": "PENDING",
                                    "transaction_date": "2018-06-29",
                                    "transaction_location_pid": "1",
                                  },
                                }
                              }
                            />
                          </div>
                        </GridColumn>
                        <GridColumn
                          computer={3}
                          largeScreen={3}
                        >
                          <div
                            className="three wide computer three wide large screen column"
                          />
                        </GridColumn>
                        <GridColumn
                          computer={2}
                          largeScreen={2}
                          textAlign="right"
                        >
                          <div
                            className="right aligned two wide computer two wide large screen column"
                          >
                            <Link
                              to="/backoffice/documents/docid-1"
                            >
                              <LinkAnchor
                                href="/backoffice/documents/docid-1"
                                navigate={[Function]}
                              >
                                <a
                                  href="/backoffice/documents/docid-1"
                                  onClick={[Function]}
                                >
                                  Document 
                                  <DocumentIcon>
                                    <Icon
                                      as="i"
                                      name="book"
                                    >
                                      <i
                                        aria-hidden="true"
                                        className="book icon"
                                        onClick={[Function]}
                                      />
                                    </Icon>
                                  </DocumentIcon>
                                </a>
                              </LinkAnchor>
                            </Link>
                            <br />
                          </div>
                        </GridColumn>
                      </div>
                    </Grid>
                  </div>
                </ItemContent>
                <div
                  className="pid-field"
                >
                  #
                  loanid-1
                </div>
              </div>
            </Item>
          </LoanListEntry>
        </div>
      </ItemGroup>
    </LoanList>
  </Router>
</BrowserRouter>
`;

exports[`LoanList tests should show empty message 1`] = `<SearchEmptyResults />`;
